Add two classes for HasHDF/HasDict compat

pmrv opened the pull request:
and outline transition path in docstrings there.

pmrv committed:
Add two classes for HasHDF/HasDict compat

and outline transition path in docstrings there.

pmrv committed:
Make GenericJob compatible with HasDict again

pmrv committed:
Add explicit error in DummyHDFio when project is not set

pmrv committed:
Use HasDictfromHDF in DataContainer

pmrv commented:
@jan-janssen and I discussed something like this last week.    The problem statement is essentially: We want everything as to_dict, but don't want to convert all to_hdf at once.  Some code also still expects to be able to call to_hdf and again we don't want to change those all at one as well.  So this introduces two mixins that can translate implementers of HasHDF and HasDict between each other. If we use them we can then slowly fade out to_hdf implementation by manually converting them to HasHDF over time.  Once all objects are converted, we deprecate HasHDFfromDict.to_hdf to show us what we need to update in the surrounding code, and when that is done we remove both classes again.

Tests will need to follow and it's not yet used by any objects, but I want to discuss this plan and the design first.

pmrv committed:
Adapt GenericParameters

pmrv commented on pyiron_base/interfaces/has_dict.py:
```diff
@@ -37,4 +53,61 @@ def _type_to_dict(self):
+    """
+
+    def _from_hdf(self, hdf, version=None):
+        entries = self._get_hdf_entries()
```
```suggestion
```

pmrv commented on pyiron_base/interfaces/has_dict.py:
```diff
@@ -37,4 +53,60 @@ def _type_to_dict(self):
+    Implements HasDict in terms of HasHDF.
+
+    This class is intended for "old-style" objects that should be able to be
+    used as children for objects that already implements HasDict and expect
```
```suggestion
    used as children for objects that already implement HasDict and expect
```

jan-janssen commented on pyiron_base/storage/parameters.py:
```diff
@@ -500,14 +500,13 @@ def define_blocks(self, block_dict):
         """
-        data_dict = self._type_to_dict()
         data_dict["data_dict"] = self._dataset
         return data_dict
```
```suggestion
        return {"data_dict": self._dataset}
```

jan-janssen commented on pyiron_base/storage/parameters.py:
```diff
@@ -500,14 +500,13 @@ def define_blocks(self, block_dict):
         Returns:
             dict: GenericParameters object as a dictionary
         """
-        data_dict = self._type_to_dict()
```
Removing this line causes the tests to fail

pmrv commented:
```
Error: The action 'Test' has timed out after 5 minutes.
```

pmrv commented on pyiron_base/interfaces/has_dict.py:
```diff
@@ -25,10 +41,10 @@ def from_dict(self, obj_dict: dict, version: str = None):
         pass
 
-    def _type_to_dict(self):
+    def to_dict(self):
```
```suggestion
    @HasDict.to_dict
    def to_dict(self):
```
What about this so people don't have to put `_...?`

pmrv replied:
Do literature search what is common in the community?  Some reference to OOP best practices?

liamhuber replied:
I poked around for this, but I found nothing conclusive. The best I got was [this SO thread](https://stackoverflow.com/questions/30599316/using-abstract-private-methods-as-an-interface) from eight years ago; they lay out the same choices we're already aware of (OP's default was to do it the way you have here, if that's anything), but the question is left unanswered as "well, that's opinion". Personally, I find the `@abstractmethod` flag on the private method of the same name the clearest...but I agree it's opinion.

pmrv replied:
I would like to stick to the current convention then.  Since this is a private class used internally anyway, it should be fairly easy to change in the future anyway, in case we decide differently.

pmrv committed:
Add tests for HasDictFromHDF

I had to add write/read_dict_to_hdf to the DummyHDFio to make it work.

pmrv commented:
I've added a test in form of the `DataContainer` implementation of `HasDictFromHDF` and adapted `DummyHDFio` to work with this.

@jan-janssen I'd prefer to merge this in before proceeding with your `to_dict` series of PRs.   Changes needed there to use this would hopefully be minimal.

I've noticed as well when running the tests locally that there is one location where we create a job with `run_mode=="manual"` that hangs for a long time before the tests proceed.  This may be the issue of the timeout problems we are seeing with the CI, but I couldn't locate the test case yet.  I have a feeling this may be related to the `files` changes with the decompression that we did recently.  I have also seen a bunch unrelated test failures locally.

jan-janssen commented:
> @jan-janssen I'd prefer to merge this in before proceeding with your `to_dict` series of PRs. Changes needed there to use this would hopefully be minimal.

The only pull requests which use `HasDict` are https://github.com/pyiron/pyiron_base/pull/1380 and https://github.com/pyiron/pyiron_base/pull/1377 . As those are required for https://github.com/pyiron/pyiron_atomistics/pull/1356 I would prefer to merge those first before we change the interface.

pmrv commented:
The reason for the unit tests hanging is this line

https://github.com/pyiron/pyiron_base/blob/0bdde9b73eb32f85688808898b208f716b606689/tests/job/test_worker.py#L26

pmrv commented:
> The reason for the unit tests hanging is this line
> 
> https://github.com/pyiron/pyiron_base/blob/0bdde9b73eb32f85688808898b208f716b606689/tests/job/test_worker.py#L26

What I don't understand is why this is suddenly a problem as this  PR doesn't touch worker code at all.

pmrv commented:
Seems this only breaks, once I touched `GenericParameters`.

pmrv commented:
@jan-janssen I can't seem to fix this quickly today, so go ahead with the other PRs.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pmrv committed:
Adapt Server to new _to_dict

pmrv committed:
Quick refactor of DataContainer.to_builtin

pmrv committed:
Use HasDict to break down values in to_builtin

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pmrv committed:
Adapt Executable to new HasDict

pmrv committed:
Respect group name

pmrv committed:
Implement HasDict directly for DataContainer

pmrv committed:
Implement HasDict directly for DataContainer

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pmrv committed:
Adapt Executable.from_dict to DataContainer

pmrv committed:
Recursive to/from_dict

Adds a create_from_dict function that takes the output of
HasDict.to_dict and turns it into a live object, similar to our earlier
to_object() method on ProjectHDFio.
Adds an instantiate class method to HasDict to support this, this allows
HasDictfromHDF to work and will be useful for dataclasses in the future.
HasDict.to_dict now goes over the contents of what is returned from
_to_dict and automatically converts any HasDict/HasHDF objects it finds.
I haven't used this in downstream code yet to keep the change small, but
in principle this will allow GenericJob/DataContainer to stop calling
to_dict on their children explicitly and let the generic interface
handle it.

The rest of the changes are renaming everything to _from_dict/_to_dict
and normalizing the argument name to obj_dict.

pmrv committed:
Redefine HasDict._type_to_dict so that GenericJob can overload it

pmrv committed:
Make _type_to_dict compatible across the various Has* classes

pmrv committed:
Include DICT_VERSION in internal nodes and restore read_only

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jan-janssen commented on pyiron_base/jobs/job/extension/executable.py:
```diff
@@ -203,13 +203,10 @@ def executable_path(self, new_path):
+    def _to_dict(self):
+        full_dict = self.storage.to_dict()
+        data = full_dict.pop("data")
+        return {"executable": {**full_dict, **data}}
```
In principle, I like the idea with the `DummyHDFio`. But for classes like the `Executable`, I am not even sure if they need a `DataContainer`. I think the goal is to get to `pickle`-able classes in the future, with the `to_dict()` and `from_dict()` method being part of the transition. So I feel switching to `_to_dict()` here goes in the wrong direction.

pmrv replied:
The DummyHDFio is not used for the executable or the data container, so I'm not sure what you mean.  I agree that in the future `Executable` could be a simple dataclass or so, but again here I just modified it, so it conforms to the new _to_dict.

jan-janssen commented on pyiron_base/jobs/job/extension/server/generic.py:
```diff
@@ -602,24 +602,22 @@ def view_queues():
+            "run_time": self.run_time,
+            "memory_limit": self.memory_limit,
+            "accept_crash": self.accept_crash,
+        }
```
In analogy to the `Executable` class, does the `Server` class need the transition from `to_dict()` to `_to_dict()` or can we modify the class in a way that the `__getstate__()` returns the same dictionary as `to_dict()`. Again my focus would be on simplifying the class hierarchy.

pmrv replied:
Both Server and Executable derived from `HasDict` before this PR.  I'm ok with changing this in the future, but here I just did the minimal changes to keep existing sub classes working with the changes I did to `HasDict` (which is just adding the type info automatically to the results of to_dict).

jan-janssen replied:
I add this to the bucket list. I would like to replace both with data classes sooner rather than later then we can potentially also reuse them in the functional approach.

jan-janssen reviewed (approved):
Looks good to me

jan-janssen commented:
@pmrv As https://github.com/pyiron/pyiron_base/pull/1578 and  https://github.com/pyiron/pyiron_base/pull/1580 are now merged - can you update this pull request, so we can include it in the next release? I think the three pull requests together streamline the `to_dict()` and `from_dict()` interface, so I would like to release them as part of `pyiron_base=0.10.0`.

pmrv commented:
contrib tests fail because its dependencies cannot currently be installed concurrently with the ones from atomistics.

pmrv commented:
This is in a good state now, but I'm still trying to add recursive to_dict/from_dict (in a separate PR).

pmrv committed:
Fix typos

jan-janssen commented:
@pmrv Can this be merged? Or should we wait for ta new `h5io_browser` version to include https://github.com/h5io/h5io_browser/pull/61 ?

jan-janssen commented on pyiron_base/jobs/job/extension/server/generic.py:
```diff
@@ -561,10 +561,9 @@ def view_queues():
+    def _to_dict(self):
         self._data.run_mode = self._run_mode.mode
-        server_dict.update(asdict(self._data))
+        return asdict(self._data)
         return server_dict
```
```suggestion
        return asdict(self._data)
```

pmrv commented:
Both this and #1602 dont need the h5io fix, thats just something I found while trying to get the jobs read their dicts from HDF recursively.

pmrv committed:
Add docstrings and rearrange method order

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pmrv committed:
Merge pull request #1602 from pyiron/hasdictrec

[minor] Recursive to/from_dict
